rust2/core/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+    ///    of predefined statistic types, TODO (change this when statistics enum is added )
+    ///     see \ref adbc-table-statistics. To get
+    ///    driver-specific statistic names, use [Connection::get_statistic_names].
+    /// 3. If true, then the value is approximate or best-effort.
+    ///
+    /// VALUE_SCHEMA is a dense union with members:
+    ///
+    /// | Field Name               | Field Type                       |
+    /// |--------------------------|----------------------------------|
+    /// | int64                    | int64                            |
+    /// | uint64                   | uint64                           |
+    /// | float64                  | float64                          |
+    /// | binary                   | binary                           |
+    ///
+    /// # Since
+    ///
+    /// ADBC API revision 1.1.0
+    fn get_statistics(
+        &self,
+        catalog: Option<&str>,
+        db_schema: Option<&str>,
+        table_name: Option<&str>,
+        approximate: bool,
+    ) -> Result<impl RecordBatchReader + Send>;
+
+    /// Commit any pending transactions. Only used if autocommit is disabled.
+    ///
+    /// Behavior is undefined if this is mixed with SQL transaction statements.
+    fn commit(&self) -> Result<()>;
+
+    /// Roll back any pending transactions. Only used if autocommit is disabled.
+    ///
+    /// Behavior is undefined if this is mixed with SQL transaction statements.
+    fn rollback(&self) -> Result<()>;
+
+    /// Retrieve a given partition of data.
+    ///
+    /// A partition can be retrieved from [Statement::execute_partitions].
+    ///
+    /// # Arguments
+    ///
+    /// - `partition` - The partition descriptor.
+    fn read_partition(&self, partition: &[u8]) -> Result<impl RecordBatchReader + Send>;
+}
+
+/// A handle to an ADBC statement.
+///
+/// A statement is a container for all state needed to execute a database query,
+/// such as the query itself, parameters for prepared statements, driver
+/// parameters, etc.
+///
+/// Statements may represent queries or prepared statements.
+///
+/// Statements may be used multiple times and can be reconfigured
+/// (e.g. they can be reused to execute multiple different queries).
+/// However, executing a statement (and changing certain other state)
+/// will invalidate result sets obtained prior to that execution.
+///
+/// Multiple statements may be created from a single connection.
+/// However, the driver may block or error if they are used concurrently
+/// (whether from a single thread or multiple threads).
+pub trait Statement: Optionable<Option = OptionStatement> {
+    /// Bind Arrow data. This can be used for bulk inserts or prepared
+    /// statements.
+    fn bind(&self, batch: RecordBatch) -> Result<()>;
+
+    /// Bind Arrow data. This can be used for bulk inserts or prepared
+    /// statements.
+    fn bind_stream(&self, reader: Box<dyn RecordBatchReader + Send>) -> Result<()>;
```

Comment:
That would mean that an implementation of `Statement` will have to define which impl of `RecordBatchReader` it will use here.

I would rather suggest making bind_stream generic with `R: RecordBatchReader + Send`.

Reply:
Yes, we can make this function generic but the bound would need to be `R: RecordBatchReader + Send + 'static`.

To fully grasp the reason behind this, we need to consider the driver manager, in particular this line: https://github.com/alexandreyc/adbc-rs/blob/24990185cc5ab37b35ce2e8157fbb16f5608c19d/core/src/driver_manager.rs#L973

The reason I choose this design is because `FFI_ArrowArrayStream::new` takes an `Box<dyn RecordBatchReader + Send>`. We can switch to a generic but we need to box it before giving it to `FFI_ArrowArrayStream::new` and in that case the compiler complains that the generic does not live long enough.

I'm open to reconsider this (and https://github.com/apache/arrow-adbc/pull/1725#discussion_r1567748242) but I think you should first read the code for the driver manager and exporter. So maybe I could add a TODO and rediscuss it in follow-up PRs?

Reply:
+1 for a TODO for now.